src/librustc_mir/borrow_check.rs

```diff
@@ -846,6 +850,199 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
             }
         }
     }
+
+    /// Check the permissions for the given lvalue and read or write kind
+    fn check_access_permissions(&self, (lvalue, span): (&Lvalue<'tcx>, Span), kind: ReadOrWrite) {
+        match kind {
+            Write(WriteKind::MutableBorrow(BorrowKind::Unique))
```

Comment:
Hmm I forgot about unique borrows when writing up those rules. They have a different set of conditions than mutable borrows. I think that this actually lines up perfectly with what I called `can_host_mutable_ref`, which we could perhaps just call `is_unique`. I forget why I changed it from that name. We should just call it `is_unique`, and then check that `Write(WriteKind::MutableBorrow(BorrowKind::Unique))` requires `is_unique`.